[ADD] new module  pos_store_draft_order

legalsylvain opened the pull request:
Hi, 

I propose a module 'pos_slate' to manage 'slate' in a shop. IE : a custom comes and  takes products, but don't pay for any reason. At the end of the session, without this module, the cashier can not close him session. With that module he can. 
- The draft PoS Order will be reaffected to the new session;
- Partial paid PoS Order will prevent to close session;
- Only users with a pos.config with new setting 'allow_slate' can do that;

Please make only a functional review and say me if you're interested by the module. 
If yes, I'll improve this PR. (OCA convention / new api)
Regards.

legalsylvain committed:
[ADD] new module pos_slate;

ajite commented on pos_slate/__openerp__.py:
```diff
@@ -0,0 +1,75 @@
+
+{
+    'name': 'Point Of Sale - Keep Draft Orders ',
+    'version': '2.1',
```
Please check the guidelines on how to define the module version number.
https://github.com/OCA/maintainer-tools/blob/master/CONTRIBUTING.md#version-numbers

ajite commented:
Hi Sylvain,

Technically it looks good to me. Since it's an Odoo 8 menu, It would have been good to use the new API.

legalsylvain commented:
Hi @ajite 
Thanks for your review !
As said in my 1er comment, If people are interested by the functionalities of this module, I'll refactor with the new API.

ajite commented on pos_slate/model/pos_session.py:
```diff
@@ -0,0 +1,67 @@
+                                po.name)))
+                # remove session id on the current PoS if it is in draft mode
+                if po.state == 'draft' and ps.config_id.allow_slate:
+                    po_obj.write(cr, uid, po.id, {
```
Since you are always writing the same thing on the objects, I would recommend declaring a po_ids list before the

```
for ps in self.browse(cr, uid, ids, context=context):
```

Then

```
if po.state == 'draft' and ps.config_id.allow_slate:
    po_ids.append(po.id)
```

And to use the PO object write method after the loop

```
po_obj.write(cr, uid, po_ids, {'session_id': None}, context=context)
```

It will reduce the number of writes.

ajite commented:
@legalsylvain: I know you said no technical review but your code looks already very clean to me :)! The module seems interesting, so I am sure @elicoidal would be glad to give his functional point of view.

elicoidal commented:
I will have a look and test functionally

legalsylvain commented:
Note : 
with Odoo / PoS, **pos_slate** allows to have PoS Order created from _back-office_ in slate;
If you use **pos_slate** + **pos_order_load** (#42), you'll have the possibility to set PoS Order created from _front-office_ in slate;

elicoidal commented:
I am confused by the name of the module. I couldnot find any specific term explaining it.
Could you help me here: is slate a specific business term?

legalsylvain commented:
Hi,
Slate is (for me) a non paid order during a long time. It's not very specific business but it is pretty common in restaurants or in bars, when the owner of the bars knows some customers and let them to pay at the end of the month, of the year, (or never... but this is another story)
For french speaker, this is "ardoise". I google-translated the terme and it gave me "slate".

I'm really not very fluent in english, so propositions are welcome. 

Thanks a lot for your review.

elicoidal commented on pos_slate/__openerp__.py:
```diff
@@ -0,0 +1,75 @@
+
+    * A new field 'allow_slate' is available in PoS Config Model, to allow
+      or block the user to let Orders in slate;
+
```
add Try me on runbot

elicoidal commented on pos_slate/tests/tests.py:
```diff
@@ -0,0 +1,120 @@
+
+        self.assertEquals(
+            ps1.state, 'closed',
+            """Draft Orders without Payment must not block the closing
```
"Unpaid Draft Orders must not block the closing"

elicoidal commented on pos_slate/model/pos_session.py:
```diff
@@ -0,0 +1,67 @@
+    def wkf_action_closing_control(self, cr, uid, ids, context=None):
+        """Remove all PoS Orders in 'draft' to the sessions we want
+        to close.
+        Check if there is Partial Paid Orders"""
```
Check if there is Partial Paid Orders =>Check if there is any Partial Paid Orders

elicoidal commented on pos_slate/model/pos_session.py:
```diff
@@ -0,0 +1,67 @@
+                if po.is_partial_paid:
+                    raise Warning(
+                        _("You cannot confirm this session, because '%s'"
+                            " is in a 'draft' state with payments.\n\n"
```
is in a 'draft' state with payments
=> is still in 'draft' state with associated payments

elicoidal commented on pos_slate/model/pos_session.py:
```diff
@@ -0,0 +1,67 @@
+                    raise Warning(
+                        _("You cannot confirm this session, because '%s'"
+                            " is in a 'draft' state with payments.\n\n"
+                            "Please finish to pay this Order." % (
```
Please finish to pay this Order.
=> Please finish to pay this Order first.

elicoidal commented:
Not sure how to test the feature as far as I understood. 
I went to runbot with existing demo data (main is allow_slate) and tried to have a session with unpaid orders but they are not saved over sessions as claimed or at least it seems I am missing something.
Any guideline on how to test it?

legalsylvain committed:
[REF] change file headers according to the new name of the module;

legalsylvain committed:
[FIX] bad terminology;

legalsylvain committed:
[REF] Oca convention;

legalsylvain committed:
[REF] remove useless headers in init files, to reduce the size of the cloud;

legalsylvain committed:
[REF] use new API;
[REF] move documentation from openerp manifest file to readme rst file;
[ADD] screenshot images and some sentences to describe the module;

legalsylvain committed:
[FIX] bad folder for images;

legalsylvain committed:
[FIX] readme file;

legalsylvain commented:
Hi @elicoidal . I updated the manifest. 
https://github.com/grap/pos/tree/8.0_ADD_pos_slate/pos_slate

Without pos_order_load module, this module allow user that use PoS (back-office), to create draft orders and to close the session.
To do that, just open a session, go back in back office, create some orders, mark some orders as paid and let other unpaid, and close the session.

(With pos_order_load #42 , it is possible to do the same with the front office version of the PoS.)

legalsylvain committed:
[REF] improve description;

legalsylvain committed:
[FIX] description;

legalsylvain commented:
Hi @elicoidal, @ajite. if slate is not a good word, what about a name like 'pos_allow_draft_orders' for this module ?
Kind regards.

elicoidal commented:
Looks better from my window but we could ask an anglophone.
Sorry for my absence. Still need to test it but basically fully booked...

Sent from my Samsung device

-------- Original message --------
From: Sylvain LE GAL notifications@github.com 
Date: 2015/09/22  16:27  (GMT+08:00) 
To: OCA/pos pos@noreply.github.com 
Cc: Eric Caudal eric.caudal@elico-corp.com 
Subject: Re: [pos] [ADD] new module pos_slate; (#49) 

Hi @elicoidal, @ajite. if slate is not a good word, what about a name like 'pos_allow_draft_orders' for this module ?
Kind regards.

—
Reply to this email directly or view it on GitHub.

ajite commented:
@legalsylvain : The name sounds fine to me, but then English is not my mother tongue.

liebana commented:
@pedrobaeza is the good guy with the module names, but your proposal seems ok to me. Is then everything ready to review @legalsylvain ?

pedrobaeza commented:
My bet is for `pos_store_draft_order` (conventions tell about singular on all).

legalsylvain commented:
@pedrobaeza : Thanks !
I'll make the changes.
regards.

legalsylvain committed:
[REF] rename module and improve tests;

legalsylvain committed:
[FIX] flake8

legalsylvain commented:
@pedrobaeza : done. 
@ajite, @elicoidal, @liebana : ready for review. I hope that the new terms will be more understandable.

JGarcia-Panach commented:
Dear @legalsylvain, thank you very much for all your hard work.
I have tested your two modules. I write my tests bellow.

**Testing pos_state_draft_order**

**1.- I Install pos_state_draf_order**

**2.- I Activate - Allow to Store Draft Orders in POS configuration**
(I have a previous pos_order from a sale in the POS frontend (order TPV/0001 - session POS/01)
![1 - antes de crear la orden](https://cloud.githubusercontent.com/assets/13234598/10634109/c5fefa4c-77ef-11e5-9717-6479f566beb1.JPG)

**3.- I Open a new session (POS/02)**

**4.- I Create an order from POS backend (order TPV/0002)**
![2 - tras crear la orden en la sesion 2](https://cloud.githubusercontent.com/assets/13234598/10634136/f9adbaea-77ef-11e5-9dd5-91ec90fc685f.JPG)

**5.- I Close session (POS/02)**
The order remains in draft state - Now the order has no session asigned (but sale ref. is TPV/0002)

**6.- I open another session (POS/03)**

**7.- I make another sale from the frontend**
![3 - tras hacer venta en tpv3](https://cloud.githubusercontent.com/assets/13234598/10634206/753531a2-77f0-11e5-8938-6629abc70699.JPG)

Now the pos_order has session POS/03 asigned (and still ref TPV/0002) **works fine for me**
![4 - now the pos_order has session 3 asigned](https://cloud.githubusercontent.com/assets/13234598/10634246/c5d190ba-77f0-11e5-801b-612aba6ac611.JPG)

**7.- I Close session (POS/03)**
This is the accounting from my three sessions
![5 - imagen5](https://cloud.githubusercontent.com/assets/13234598/10634267/e530545a-77f0-11e5-9817-27d3594618f1.JPG)

**8.- Now without opening another session, I try to pay my pos_order TPV/0002**
But I'm not allowed, because I don't have an active session - **Perfect for me again**

**9.- I open a new session POS/04 and pay the draft order**

**10.- I close the session POS/04**
Here are the pos_orders
![6 - here the orders](https://cloud.githubusercontent.com/assets/13234598/10634316/40004354-77f1-11e5-8b13-dcda98513b73.JPG)

And here the accounting
![7 - accounting](https://cloud.githubusercontent.com/assets/13234598/10634325/4a79d674-77f1-11e5-9e93-67d0c51e384c.JPG)

**ALL PERFECT**

**Testing pos_order_load**

**1.- I open a new session (POS/05)**
And in frontend I save a pos_order (order TPV/0007)
![8 - before saving](https://cloud.githubusercontent.com/assets/13234598/10634413/e4532322-77f1-11e5-98ec-cc49c4b180e8.JPG)

![10 - imagen10](https://cloud.githubusercontent.com/assets/13234598/10634417/ed24b498-77f1-11e5-86a6-3a8272966c0a.JPG)

**2.- Without closing the session (POS/05) and still in the frontend** 
Each time I load my order, validate, and save it again, it increase the order's name in one (in the image you can see the order after load, validate  and save for three times) Now the order's name is TPV/0010)
**The sequence I have followed is: LOAD - SELECT ORDER - VALIDATE - SAVE**
![11 - imagen con sale order 10](https://cloud.githubusercontent.com/assets/13234598/10634455/2cfae754-77f2-11e5-8f84-9c030fb127ef.JPG)

If I push Load Draft Order button and push validate without selecting an order, this is the error message. **Maybe it would be better another error  message** just IMO
![12 - error](https://cloud.githubusercontent.com/assets/13234598/10634521/ae32cf30-77f2-11e5-9a23-f7d7fcbfa564.JPG)

**3.- The problem is when I follow this other sequence:**
**LOAD - SELET ORDER - SAVE** (without pressing validate button)
**Another pos_order is created every time you press save button**
I know it is a bad performance by my side, but I don't know if it should be controlled by the module.
![14 - varias ordenes creadas](https://cloud.githubusercontent.com/assets/13234598/10634569/166e568c-77f3-11e5-8bef-b88d001916f4.JPG)

You also can add more products to the order and save again with no conflicts of any kind if you follow the correct sequence.

I hope this could help. And if anyone consider my test comments should be shorter please, let me know. I am learning how to contribute with the community.

Regards

liebana commented:
Your comments are perfect @JGarcia-Panach !

JGarcia-Panach commented:
Thanks man!! @liebana

legalsylvain commented:
Hi @JGarcia-Panach, 

Thanks a lot for your complete review and sorry for the late. If I understand correctly, this module is OK, but the second one (pos_load_order) (https://github.com/OCA/pos/pull/42) has a problem with sequences.
So are you ok to merge this PR and make some fixes on #42.

Kind regards.

legalsylvain commented:
Up !  some reviews for this PR but no thumbs up. Any reason ? 
regards !

pedrobaeza commented:
@JGarcia-Panach, can you answer @legalsylvain question?

JGarcia-Panach commented:
Sorry @legalsylvain and @pedrobaeza, I need to test it again. Let me do it next week and I let you now.
Regards

JGarcia-Panach commented:
@legalsylvain, tested on runbot and it works perfectly, this time I've tested just this module.

I've opened a session and I've sold several products in the frontend and created a new pos order in the backend (and I haven't paid it).
Iv'e closed the session with the order in draft state and  I have paid it in the next session.
Everything worked fine.

However, to have the full package, it would be great if this module also would decreased stock when the client takes the product and doesn't pay for any reason.

I will try to test both modules together within the next week.
(I have a question about this, can I test them together on runbot?)

But for this module, you have my 👍 
Thanks for your hard work!

legalsylvain commented:
Hi @JGarcia-Panach. Thanks for your review ! About the feature you ask, not sure it is simple. The POS behaviour is to make "all include". the sale is done when you pay. Before, the pos.order is in a draft mode. So it can be fully cancel, deleted, ... So stock move can not be done. (or require a lot of changes, adding a intermediary state 'sold_not_paid' ...)
If you want such feature, please refer better to pos.order to sale.order module, that could feat your need correctly.

legalsylvain commented:
I know that this PR has only one :+1: by @JGarcia-Panach  but could we consider to accept it ?
CC : @elicoidal , @pedrobaeza.

Allways very hard to have a lot of reviewers in OCA / pos repository.

Regards.

elicoidal commented:
@legalsylvain I will try to have a look in the coming days.

rafaelbn reviewed (approved):
Thanks @legalsylvain sorry for this late!
